### PR TITLE
Cleanup and refactoring of TrackDAO: Adding tracks [rebased and consolidated]

### PR DESCRIPTION
**Pull request by uklotzde**
TrackDAO has too many responsibilities, a multitude of hardly testable code paths, and subtle assumptions about how various member functions are supposed to interact.

As a first attempt this PR tries to straighten the code paths when adding single/multiple tracks, which is needed for the introduction of TrackCache.

### CODE REVIEW

File: src/library/dao/trackdao.cpp [hunk 1]
```diff
@@ -37,16 +36,14 @@ QMutex TrackDAO::m_sTracksMutex;
 
 enum { UndefinedRecordIndex = -2 };
 
-TrackCacheItem::TrackCacheItem(TrackPointer pTrack)
-        : m_pTrack(pTrack) {
-    DEBUG_ASSERT(m_pTrack);
+RecentTrackCacheItem::RecentTrackCacheItem(TrackPointer pTrack)
+        : m_pTrack(std::move(pTrack)) {
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: QSharedPointer has no move constructor, so std::move() is redundant. Instead we can pass the pTrack by const reference.

File: src/library/dao/trackdao.cpp [hunk 2]
```diff
@@ -37,16 +36,14 @@ QMutex TrackDAO::m_sTracksMutex;
...
+    DEBUG_ASSERT(!m_pTrack.isNull());
 }
 
-TrackCacheItem::~TrackCacheItem() {
-    DEBUG_ASSERT_AND_HANDLE(!m_pTrack.isNull()) {
-        return;
-    }
-    // qDebug() << "~TrackCacheItem" << m_pTrack << "ID"
+RecentTrackCacheItem::~RecentTrackCacheItem() {
+    DEBUG_ASSERT(!m_pTrack.isNull());
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Why did you remove .._AND_HANDLE? Can't we remove the whole statement, since we check it in the constructor as well?

File: src/library/dao/trackdao.cpp [hunk 3]
```diff
@@ -152,7 +149,7 @@ void TrackDAO::finish() {
     // And mark the corresponding tracks in track_locations in the deleted
     // directories as deleted.
     // TODO(XXX) This doesn't handle sub-directories of deleted directories.
-    foreach (QString dir, deletedHashDirs) {
+    for (const QString& dir: deletedHashDirs) {
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Here we have decided to use "auto"

File: src/library/dao/trackdao.cpp [hunk 4]
```diff
@@ -523,32 +520,35 @@ void TrackDAO::addTracksFinish(bool rollback) {
...
+                "been prepared. Skipping track"
+                << pTrack->getLocation();
+        return TrackId();
     }
 
+    qDebug() << "TrackDAO: Adding track"
+            << pTrack->getLocation();
+
     int trackLocationId = -1;
     TrackId trackId;
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: I think you can reuse this invalid TrackId above and below.

File: src/library/dao/trackdao.cpp [hunk 5]
```diff
@@ -599,14 +599,14 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
                 }
             }
             // Regardless of whether we unremoved this track or not -- it's
-            // already in the library and so we need to skip it. Set the track's
-            // trackId so the caller can know it. TODO(XXX) this is a little
-            // weird because the track has whatever metadata the caller supplied
-            // and that metadata may differ from what is already in the
-            // database. I'm ignoring this corner case. rryan 10/2011
-            pTrack->setId(trackId);
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Where is this call gone?

File: src/library/dao/trackdao.cpp [hunk 6]
```diff
@@ -628,49 +628,53 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
...
-    m_tracksAddedSet.insert(trackId);
-    return true;
+    if (trackId.isValid()) {
+        pTrack->setId(trackId);
+    } else {
+        qWarning() << "TrackDAO::addTracksAdd:"
+                << "Failed to add track to database"
+                << pTrack->getLocation();
+    }
+    DEBUG_ASSERT(pTrack->getId() == trackId);
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: this Assert here is hard to understand some comments about the possible cases would be useful.

File: src/library/dao/trackdao.cpp [hunk 7]
```diff
@@ -682,41 +686,39 @@ QList<TrackId> TrackDAO::addTracks(const QList<QFileInfo>& fileInfoList,
...
+        return QList<TrackId>();
     }
 
-    // All all the track paths to this database.
+    // Add all the track paths to this database.
     query.prepare("INSERT INTO playlist_import (add_index, location) "
                   "VALUES (:add_index, :location)");
     int index = 0;
-    foreach (const QFileInfo& rFileInfo, fileInfoList) {
+    for (const QFileInfo& fileInfo: fileInfoList) {
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: auto

File: src/library/dao/trackdao.cpp [hunk 8]
```diff
@@ -733,25 +735,17 @@ QList<TrackId> TrackDAO::addTracks(const QList<QFileInfo>& fileInfoList,
         LOG_FAILED_QUERY(query);
     }
     const int addIndexColumn = query.record().indexOf("add_index");
-    const int locationColumn = query.record().indexOf("location");
     while (query.next()) {
         int addIndex = query.value(addIndexColumn).toInt();
-        QString filePath = query.value(locationColumn).toString();
-        const QFileInfo& fileInfo = fileInfoList.at(addIndex);
-        TrackPointer pTrack(new TrackInfoObject(fileInfo));
-        SoundSourceProxy(pTrack).loadTrackMetadata();
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Where do we load now the Metadata?

File: src/library/dao/trackdao.cpp [hunk 9]
```diff
@@ -1361,38 +1365,30 @@ TrackPointer TrackDAO::getTrackFromDB(TrackId trackId) const {
...
     // If the track is dirty send dirty notifications after we inserted
     // it in the cache. BaseTrackCache cares about dirty notifications
-    // and the setDirty call above happens before we connect to the
+    // and the markDirty() call above happens before we connect to the
     // track's signals.
-    if (shouldDirty) {
+    if (isTrackDirty) {
         emit(trackDirty(trackId));
+    } else {
+        emit(trackClean(trackId));
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Why we need this signal here?

File: src/library/dao/trackdao.cpp [hunk 10]
```diff
@@ -628,49 +628,53 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
...
+                << "Unsupported file type"
+                << fileInfo.absoluteFilePath();
+        return TrackId();
     }
+    SoundSourceProxy(pTrack).loadTrackMetadata();
 
-    addTracksPrepare();
-    addTracksAdd(pTrack, unremove);
-    addTracksFinish(false);
+    const TrackId trackId(addTracksAdd(pTrack, unremove));
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Is this a recursive call?

Comment (daschuer) [inline src/library/dao/trackdao.cpp]: No, It is an overload function. We should rename one.

File: src/library/dao/trackdao.cpp (hunk 8 above)
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Got it. I was confused because the overload of addTracksAdd()

File: src/library/dao/trackdao.cpp [hunk 11]
```diff
@@ -1361,38 +1365,30 @@ TrackPointer TrackDAO::getTrackFromDB(TrackId trackId) const {
...
+    if (isTrackDirty) {
         emit(trackDirty(trackId));
+    } else {
+        emit(trackClean(trackId));
     }
 
-    // If the header hasn't been parsed, parse it but only after we set the
-    // track clean and hooked it up to the track cache, because this will
-    // dirty it.
-    SoundSourceProxy(pTrack).loadTrackMetadata();
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: I do not understand why this call was here. So I am afraid I cannot tell if removing this is a good idea.
Did this always reload the track Metadata from file? Is it a regression from a previous commit? 
It should be possible to have different Metadata in Mixxx and the file. Was this broken before?

File: src/library/dao/trackdao.cpp (hunk 5 above)
Comment (uklotzde) [inline src/library/dao/trackdao.cpp]: Moved to the end of this function to avoid uncontrolled setting of the track id.

Future: Setting the track id will only be allowed from within the TrackCache 
that needs to keep its internal maps synchronized.

**Comment by daschuer** [comment]
The PR now LGTM, but it is conflicting. 
The only thing I am worrying about is the bpm lock bug. Does this have dependencies to it? 
I have not yet dived into it.

**Comment by uklotzde** [comment]
I first need to create a small follow up PR for RJ's fixes before we can 
continue to merge this PR. I also consider rebasing it on master, because I 
don't want to risk losing any of his modifications.

**Comment by daschuer** [comment]
OK, that seams to be reasonable.

File: src/library/dao/trackdao.cpp [hunk 12]
```diff
@@ -2064,11 +2092,11 @@ TrackPointer TrackDAO::getOrAddTrack(const QString& trackLocation,
                                      bool processCoverArt,
                                      bool* pAlreadyInLibrary) {
     TrackId trackId(getTrackId(trackLocation));
-    bool track_already_in_library = trackId.isValid();
+    bool trackAlreadyInLibrary = trackId.isValid();
 
     // Add Track to library -- unremove if it was previously removed.
     if (!trackId.isValid()) {
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: if (!trackAlreadyInLibrary ...

File: src/library/dao/trackdao.cpp [hunk 13]
```diff
@@ -2064,11 +2092,11 @@ TrackPointer TrackDAO::getOrAddTrack(const QString& trackLocation,
...
-        trackId = addTrack(trackLocation, true);
+        trackId = addSingleTrack(trackLocation, true);
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Here is room for optimization. We know that the track is new and we already head a TrackInfo Object inside of addSingleTrack. Can't we just add it to the cache?

File: src/library/dao/trackdao.cpp [hunk 14]
```diff
@@ -630,49 +632,69 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
...
+                << "Unsupported file type"
+                << pTrack->getLocation();
+        return TrackId();
     }
 
-    // Check that track is a supported extension.
-    if (!SoundSourceProxy::isFileSupported(pTrack->getFileInfo())) {
-        // TODO(XXX) provide some kind of error code on a per-track basis.
-        return;
+    SoundSourceProxy(pTrack).loadTrackMetadata();
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Here we load the track metadata even though we do not need it in any case. If the track is already in library it is not used. 
It is probably not a topic for this PR, but It looks like we can spit the location insert and the track insert in two functions and call the location insert before loadTrackMetadata()

**Comment by daschuer** [comment]
It was hard to read this PR a second time ....
And it still looks good to me. I have not catch significant changes from the original version, (the first review path). Are there any? 

I have left some further improvement comments. If you are finished with this PR. I would like to merge it without these bigger changes. They can go into another PR if you or an other thinks it is worth to improve it more.   

Just give a hint when I should merge this. Thank you.

**Comment by uklotzde** [comment]
I may have found a minor regression that needs to be fixed: Unremoved tracks 
are marked as clean although they shouldn't. This will not increase the 
complexity substantially.

And I will also take your comments into account regarding adding a single 
track to the library: Passing the allocated TIO as the result back to the 
caller instead of throwing it away.

File: src/library/dao/trackdao.cpp [hunk 15]
```diff
@@ -1185,6 +1221,20 @@ struct ColumnPopulator {
 
 #define ARRAYLENGTH(x) (sizeof(x) / sizeof(*x))
 
+void TrackDAO::cacheRecentTrack(
+    TrackId trackId,
+    const TrackPointer& pTrack) const {
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Please double indent here.

File: src/library/dao/trackdao.cpp [hunk 16]
```diff
@@ -630,49 +644,81 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
...
+        return TrackPointer();
     }
 
-    // Check that track is a supported extension.
-    if (!SoundSourceProxy::isFileSupported(pTrack->getFileInfo())) {
-        // TODO(XXX) provide some kind of error code on a per-track basis.
-        return;
+    // Initially load the metadata for the newly created track
+    // from the file.
+    SoundSourceProxy(pTrack).loadTrackMetadata();
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: If the track is already in library, this call can be skipped. 
Fixing this requires some more refactoring. So please just add a TODO

**Comment by daschuer** [comment]
Is the track added by addSingleTrack() added to the RecentTrackCache?

**Comment by daschuer** [comment]
I have done some random tests and it still works :-) LGTM 
Thank you!
